### DISCUSSION TITLE
Tracking CO2 Emissions

### DISCUSSION THREAD

**Discussion by x-tabdeveloping** [embeddings-benchmark/mteb#703] [Ideas]
Since we are trying to put emphasis on benchmark efficiency and would like to be transparent about the compute and environmental impact of our benchmark, it might be a good idea to track CO2 emissions for the runs we are going to conduct.
This is also something reviewers might give us comments about, as ACL is usually quite sensitive to ethical considerations.

I think adding emissions tracking with [CodeCarbon](https://mlco2.github.io/codecarbon/index.html) and adding total emissions per task per model to the results file could be a great idea. This would also give us an idea about which models are most and least polluting.

What do you think? (@isaac-chung, @KennethEnevoldsen, @imenelydiaker, @orionw, @Muennighoff )

**Comment by imenelydiaker** [comment]
I love this idea! Thanks for bringing this up 🙂 
Some reflexions and questions here:

1. I don't expect us to run more than 5-10 models for the paper. How should we select them?
2. Emissions depend on the models we run, bigger the model, bigger the carbon footprint as more hardware is needed. It also depends on the task, e.g. retrieval and reranking tasks are the most consuming in terms of hardware as very large corpuses are embedded.
3. This could be a section in the paper: _"Ethical considerations"_, we could report the carbon footprint of our evaluations for the paper. Btw, making the Clustering task faster helps reducing the carbon footprint.
4. Long term view, indeed it would be nice to have an estimation of the CO2 emission per task, it could be added to the results file along with `mean_eval_time` and `scores`.

**Reply by x-tabdeveloping** [reply]
Yeah I think estimating how much we managed to reduce CO2 emissions with some of the changes in the clustering tasks could also be a great point.

**Reply by x-tabdeveloping** [reply]
Regarding selecting the models:
I think we should probably choose popular open source models that are diverse enough in size and performance.
How many models and how are we planning to run? Do you have any info or insights on this @Muennighoff ?

**Reply by KennethEnevoldsen** [reply]
We can add it as a context manager around our tasks. Then each task can get a carbon emission similar to evaluation time.

**Reply by KennethEnevoldsen** [reply]
Task emission is then sum across task runs and the model emission sum across model.

**Reply by imenelydiaker** [reply]
@KennethEnevoldsen a suggested implementation here: https://github.com/embeddings-benchmark/mteb/pull/712

**Comment by isaac-chung** [comment]
At the model running stage, instead of reporting the emissions per task, would we want to present the _net decrease_ in emissions per task / overall as well? That might require running tasks before the speedups with emissions enabled.

**Reply by KennethEnevoldsen** [reply]
That would require us to switch to an older version of MTEB and run the benchmark there (we can probably just use the English benchmark).

Would be great to do that if someone has the time.

**Reply by mrshu** [reply]
I'd be happy to do that -- I might have some computing resources available.

Would you mind navigating me to the older version of the benchmark? A commit hash would perhaps suffice.

**Reply by KennethEnevoldsen** [reply]
anything 1.3.2 or before

**Reply by KennethEnevoldsen** [reply]
(e.g. https://github.com/embeddings-benchmark/mteb/releases/tag/1.2.0)

**Reply by mrshu** [reply]
I see, thanks @KennethEnevoldsen.

I guess I ought to port/cherry-pick @imenelydiaker's changes in #712 on top of it in order for the comparison to be fair and consistent, right?

**Reply by imenelydiaker** [reply]
Yeah you'll probably need the code of PR #712 to be consistent with future evals. Are you planning on having a specific branch for this or anything?

**Reply by mrshu** [reply]
@imenelydiaker yes, I was thinking about branching the code out. I took a stab at it in https://github.com/mrshu/mteb/commit/4e1bab4ef964291aadacf808465d54e93d3db4cc -- I'd appreciate if you could take a look (the state of the codebase was different and I had to do a couple of updates here and there).

Once the code is deemed solid enough, however, it brings a couple of further questions:
1. What models to run it on?
2. What tasks to run it on?
3. For the purposes of the reporting in the paper, would it make sense to re-run the evaluation a few times (say 3 or 5) and report both the means and standard deviations -- in other words, some sort of a box plot with error bars?

I'd appreciate any thoughts you might have around that.

Thanks!

**Reply by imenelydiaker** [reply]
Thank you @mrshu for updating the code, I left some comments on the commit.

For the questions you are asking:

1. Models: the discussion is open here https://github.com/embeddings-benchmark/mteb/discussions/705, we still didn't define which ones.
2. Tasks: good question, I have 2 options in mind although I don't know which one is better:
    - Run the whole English benchmark as in the script `run_benchmark_english.py` in the version you picked, we'll run the same tasks by replacing Clustering tasks with their fast version + downsampled tasks (Classification ones) for the sake of comparision. Then I think we'll extend it with newer datasets and compute a new global emission. 
    - Run only tasks that were converted to Fast, aka Clustering tasks, and downsampled Classification tasks both for English (you can add another language like German). Since the goal is to show that making the task fast & downsampling lowers carbon emissions.
_[For both options run only the old version of tasks, as that's what we need here]_ 
3. You mean standard deviation and mean of the carbon footprint? Yes looks great, I expect the value to change according to the 
evaluation time but not very much. Will also be nice to precise which hardware was used. But no need for plots imo, this should only be added to the **Ethical considerations** section, one paragraph with results should be enough. Maybe we'll add a small table in the appendix with mean and std.

**Reply by mrshu** [reply]
Thanks for the review and your responses @imenelydiaker! 

With regards to the models, do you think it would make sense to wait for that discussion to conclude?

On tasks, I would say  running the English benchmark might be a bit easier, so unless someone finds the second option significantly better, I guess we can go with that. I'll try to see if I can get it to run with the changes I've made above.

With regards to "you mean standard deviation and mean of the carbon footprint" -- yes, I was trying to make sure some Reviewer 2 doesn't come back with a review to the tune of _"they are computing total carbon footprint difference from a single measurement"_. A small table in the appendix makes sense -- thanks!

**Reply by imenelydiaker** [reply]
Hey @mrshu,

Should you run model on both versions of the benchmark? 
We need it to be done on the same hardware to be comparable.

I suggest the following: _Consider the carbon footprint estimation a different experiment_
- It would be cool to run at least 2 models  (a large one like sentence-t5-xxl and a small one like multilingual-e5-large)
- You'd run them on the old English benchmark and report the carbon footprint for each task
- Run the models again on the new version of some tasks of the same benchmark (it's better to keep the same number of tasks)

Wdyt? You can suggest other models I just gave these as an example

**Reply by mrshu** [reply]
Thanks @imenelydiaker! 

I already finished a run with the `average_word_embeddings_komninos` model, the results of which you can find at https://github.com/mrshu/mteb/commit/9722298ce0f2acf449f03242fc79601ee8ea8353

I also started a run with `sentence-t5-xxl` but it seems it will take hours if not days to finish.

I will certainly keep you posted about any further developments!

**Reply by mrshu** [reply]
@imenelydiaker the branch at https://github.com/embeddings-benchmark/mteb/compare/main...mrshu:mteb:mrshu/port-carbon-emissions-estimation now has results for

- `average_word_embeddings_komninos`
- `sentence-t5-xxl`
- `intfloat/multilingual-e5-large`

Would you find that enough comparison-wise? What do you think would be the best way and/or format of reporting carbon footprint based on the data we have?

**Reply by imenelydiaker** [reply]
Hey @mrshu , thank you for this! Yes it should be enough for the comparison.

For the respresentaiton, we will probably sum values accross tasks for each model. And do the same with the lite version. Finally, we'll have a barplot that should look like this:
![image](https://github.com/embeddings-benchmark/mteb/assets/33312980/32490ebe-1d0f-4936-8e79-210002c74ee0)

Although, we may want to include other modalities such as the performance or the time it took to run for better insights. Welcoming suggestions 🙂

**Reply by mrshu** [reply]
Ok, that's certainly doable.

I will need to run the MTEB-lite too for that to be doable. Would you happen to have any thoughts on what would be the best way of doing so now?

Thanks!

**Reply by KennethEnevoldsen** [reply]
Running MTEB-lite, seems to require #836, #835 (and #837). However, we can start running it on all tasks except for classification and retrieval

**Reply by mrshu** [reply]
Thanks a bunch @KennethEnevoldsen. I just wanted to quickly circle back to ask whether there might be a better way of running the MTEB-lite now.

Thanks!
